### DISCUSSION TITLE
NuxtLink Named Routes in 'to' object for dynamic route link rendering

### DISCUSSION THREAD

**Discussion by Maximus1000** [nuxt/nuxt#17026] [Questions] [answered]
Hi All, I'm having trouble getting NuxtLink to replicate Vue Router's behavior in rendering named route links.  Apologies if this has already been asked/answered (couldn't find this exact).  I'm assuming the answer is either known limitations of SSR or a simple mistake on my part.

When I render a list of blog post links using Vue's router-link (not in Nuxt),  the router will navigate to /posts/{slug}.
``<router-link :to="{ name: 'blog', params: { slug: post.slug }}">
  Some post
</router-link>
``

NuxtLink, while wrapping RouterLink, doesn't pick up on the slug in the same way. The params object is available in the child component but its properties are not detected in a way that renders a link with the noted slug.  The following will render a link to 'blog' but not add the post slug.
``<nuxt-link :to="{ name: 'blog', params: { slug: post.slug }}">
  Some Nuxt post
</router-link>
``

The path option of course does render my blog post link as expected.  This seems like my easiest and most consistent choice for a refactor of Vue's named route links wherein I would need to build up the path as opposed to define named routes in a router config.
``<nuxt-link :to="{ path: `blog/${post.slug}`, params: { slug: post.slug }}">
  with path option
</router-link>
``

I could simplify to a dynamic 'to' object, e.g. ``:to="`blog/${post.slug}`"`` and omit both named route and path but want to continue to pass params.  To be sure, I am replacing RouterLink's with NuxtLink everywhere as opposed to trying to use RouterLink in Nuxt directly (unsure if the above works using native router-link in Nuxt).  Is there something I'm missing or a better way to render links in Nuxt that matches the behavior in VueRouter for named routes?

Thanks in advance for any response and cheers to this amazing project.
-Marcus

// Correction regarding using dynamic path - still no solution:
- Vue-router warns that the nuxt-link example above with dynamic path will in fact drop the params object.
- Vue-router states that if you want params object, use a named route instead.  So, the above dynamic path is NOT a fix to this question if I want to pass params downstream.  
- There's clearly a problem here or something wrong with my NuxtLink strategy as I can either 
a) route correctly with a dynamic path but lack the ability to pass params to the child component. 
```<nuxt-link :to="{ path: `blog/${post.slug}`, params: {*ignored*}}">Link no params</nuxt-link>```
```[Vue Router warn]: Path "blog/corrupti-nulla" was passed with params but they will be ignored. Use a named route alongside params instead.```
b) Or, route with vue-router's router-link by importing vue-router directly, then use a named route strategy and get params to pass correctly.  Noting again that NuxtLink is not translating a named route with a dynamic id or slug string, where :to="{name:`blog/${slug}`}" resolves to 'blog' not 'blog/some-slug' for me.  I've also tried wrapping in ClientOnly with no success.  To be sure, slug is available and I can correctly resolve with a dynamic path as above, just not as a dynamic named route with a variable dependency.
- obviously, there are other ways to get data downstream but still feel like I'm missing something on NuxtLink.

Thx again for any help.

**Comment by amrnn90** [comment, marked as answer]
As mentioned [here](https://v3.nuxtjs.org/guide/directory-structure/pages#nested-routes)
If your pages directory structure is like this:
```
pages/
  blog/
    [slug].vue
```

Then nuxt will generate a named route that concatenates the parent and child names, so in this case: `blog-slug`
So you will need to do this instead:

```vue
 <nuxt-link :to="{ name: 'blog-slug', params: { slug: post.slug }}"> Some Nuxt post </nuxt-link> 
```

**Reply by Maximus1000** [reply]
That did it.  Thanks very much for the help.  I knew about how Nuxt creates the named routes, but for some reason it never dawned on me that the dynamic [slug] or [id] dir would also be added like you showed.  I never tested that approach and must have overlooked that bit in docs.  Makes perfect sense and worked immediately.  Thanks again!

**Reply by ITwrx** [reply]
just in case there are for future readers (new to nuxt, like me) having trouble with named routes while the docs are still in development:

```
const router = useRouter();
console.log(router.getRoutes())
```
will log the routes nuxt has auto-generated from your file structure so you can see what is going wrong.

`navigateTo({ name: 'admin-site-sections-edit-slug', params: { slug: slug } })` inside your script
or 
`<NuxtLink :to="{ name: 'admin-site-sections-edit-slug', params: { slug: site_section.slug }}">{{ site_section.title }}</NuxtLink>`
inside your template, works.

Some notes:

leave "slug" in the route name as the literal string 'slug'. I was trying to interpolate the real slug there, but the router already handles that from the param, evidently.

'site-sections' in my route name is a two word directory in my file structure, not two dirs.
IOW, `/admin/site-sections/edit-[slug].vue` . I doubt the router cares but i wasn't sure how i was supposed to name my dirs with more than one word and if that was why i was failing to get a match. Once i logged the routes i saw there was something goofy with the slug.

**Reply by Maximus1000** [reply]
Great notes and a more advanced use case.  Very helpful and thanks for taking the time!

**Comment by Zlinni** [comment]
Can you see my question? I try to make some routes dynamic  by authority without the Nuxt naming system and success; But it makes some mistake when I refresh the dynamic route. 
nuxt/nuxt#16700

**Reply by Maximus1000** [reply]
@Zlinni, I'm still getting familiar with Nuxt, unsure about the issue you noted.  You're doing some solid work it seems! I hope you figure it out and will watch for a solution such that I can learn.
